contributors/design-proposals/volume-user-configuration.md

```diff
@@ -0,0 +1,195 @@
+
+        // Optional list of values that are allowed for this label selector key.
+        AllowedValues []string
+
+        // Optional description. May contain newlines and links.
+        Description *string
+    }
+    ```
+
+  * It's responsibility of an admin to fill `sc.selectorOptions` when creating a StorageClass!
+  * We will update examples for all supported internal provisioners with selectorOptions for easier copy/paste.
+
+* `kubectl describe storageclass` will show available labels ad values in a user-friendly form.
+
+* There is no code in Kubernetes that validates PVCs with `selectorOptions` in StorageClasses. Users may add invalid selector keys or invalid values and it's up to the individual provisioner to validate the selector.
+
+## Examples
+
+### Ultimate StorageClass for AWS
+This example shows a rich StorageClass with three user-tunable parameters.
+
+```yaml
+apiVersion: storage.k8s.io/v1beta1
+kind: StorageClass
+metadata:
+  name: slow
+provisioner: kubernetes.io/aws-ebs
+parameters:
+  type: gp2
+  zones: us-east-1a, us-east-1b, us-east-1c, us-east-2a, us-east-2b, us-east-2c
+selectorOptions:
+  - key: failure-domain.beta.kubernetes.io/zone
+    allowedValues:
+        - us-east-1a
+        - us-east-1b
+        - us-east-1c
+        - us-east-2a
+        - us-east-2b
+        - us-east-2c
+    description: "Requested AWS availability zone"
+
+  - key: encryption.beta.kubernetes.io/enabled
+    defaultValue: "false"
+    allowedValues:
+      - "true"
+      - "false"
+    description: "Request volume encryption"
+
+  - key: encryption.beta.kubernetes.io/key-id
+    description: "ID of AWS encryption key"
+```
+
+kubectl output:
+```shell
+$ kubectl describe storageclass default
+Name:           default
+IsDefaultClass: Yes
+Provisioner:    kubernetes.io/aws-ebs
+SelectorOptions:
+    failure-domain.beta.kubernetes.io/zone
+      - requested AWS availability zone
+      - values:
+        - us-east-1a
+        - us-east-1b
+        - us-east-1c
+        - us-east-2a
+        - us-east-2b
+        - us-east-2c
+
+    encryption.beta.kubernetes.io/enabled
+      - request volume encryption
+      - values:
+        - true
+        - false (default)
+
+    encryption.beta.kubernetes.io/key-id
+      - ID of AWS encryption key
+
+Parameters: type=pd-standard,zones=us-east-1a, us-east-1b, us-east-1c, us-east-2a, us-east-2b, us-east-2c
+
+```
+
+* All selector keys are optional.
+* There is no default zone selected, provisioner may choose any zone if PVC selector does not require any.
+* Encryption is off by default.
+
+
+Regular user could then post a PVC to require an unencrypted PV in zone "1a", "1b" or "1c":
+
+```yaml
+kind: PersistentVolumeClaim
+apiVersion: v1
+metadata:
+  name: myclaim
+spec:
+  class: slow
+  accessModes:
+    - ReadWriteOnce
+  resources:
+    requests:
+      storage: 1500Mi
+  selector:
+    matchLabels:
+      encryption.beta.kubernetes.io/enabled: "false"
+    matchExpressions:
```

Comment:
This looks complicated to me. Probably easier to do a `AllowNone` or `AllowAll` style.